modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/RebalanceStatisticsUtils.java

```diff
@@ -0,0 +1,686 @@
+     */
+    private RebalanceStatisticsUtils() {
+        throw new RuntimeException("don't create");
+    }
+
+    /** Rebalance future statistics. */
+    static class RebalanceFutureStatistics {
+        /** Start rebalance time in mills. */
+        private final long startTime = currentTimeMillis();
+
+        /** End rebalance time in mills. */
+        private volatile long endTime = startTime;
+
+        /** First key - topic id. Second key - supplier node. */
+        private final Map<Integer, Map<ClusterNode, RebalanceMessageStatistics>> msgStats = new ConcurrentHashMap<>();
+
+        /** Is needed or not to print rebalance statistics. */
+        private final boolean printRebalanceStatistics = printRebalanceStatistics();
+
+        /**
+         * Add new message statistics.
+         * Requires to be invoked before demand message sending.
+         * This method required for {@code addReceivePartitionStatistics}.
+         * This method add new message statistics if
+         * {@link #printRebalanceStatistics} == true.
+         *
+         * @param topicId Topic id, require not null.
+         * @param supplierNode Supplier node, require not null.
+         * @see RebalanceMessageStatistics
+         * @see #addReceivePartitionStatistics(Integer, ClusterNode, GridDhtPartitionSupplyMessage)
+         */
+        public void addMessageStatistics(final Integer topicId, final ClusterNode supplierNode) {
+            assert nonNull(topicId);
+            assert nonNull(supplierNode);
+
+            if (!printRebalanceStatistics)
+                return;
+
+            msgStats.computeIfAbsent(topicId, integer -> new ConcurrentHashMap<>())
+                .put(supplierNode, new RebalanceMessageStatistics(currentTimeMillis()));
+        }
+
+        /**
+         * Add new statistics by receive message with partitions from supplier
+         * node. Require invoke {@code addMessageStatistics} before send
+         * demand message. This method add new message statistics if
+         * {@link #printRebalanceStatistics} == true.
+         *
+         * @param topicId Topic id, require not null.
+         * @param supplierNode Supplier node, require not null.
+         * @param supplyMsg Supply message, require not null.
+         * @see ReceivePartitionStatistics
+         * @see #addMessageStatistics(Integer, ClusterNode)
+         */
+        public void addReceivePartitionStatistics(
+            final Integer topicId,
+            final ClusterNode supplierNode,
+            final GridDhtPartitionSupplyMessage supplyMsg
+        ) {
+            assert nonNull(topicId);
+            assert nonNull(supplierNode);
+            assert nonNull(supplyMsg);
+
+            if (!printRebalanceStatistics)
+                return;
+
+            List<PartitionStatistics> partStats = supplyMsg.infos().entrySet().stream()
+                .map(entry -> new PartitionStatistics(entry.getKey(), entry.getValue().infos().size()))
```

Comment:
Entry value can be `null` in case of historical rebalance used. Will we get `NullPointerException` here?